### PR TITLE
core: Provide read access to commit/message log and odb

### PR DESCRIPTION
**Pull request by kim**
# Description of Changes

This is a first approximation to provide data access for Kafka-style replication. It punts on notifications of segment rotation (which could also be solved via filesystem events).

ObjectDB access is fairly crude, and will likely require it's own replication subsystem.

~~Note that this is against `cloud-next`.~~

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

### CODE REVIEW

**Comment by kim** [comment]
This will probably do, so rebased into `master`

File: crates/core/src/db/message_log.rs [hunk 1]
```diff
@@ -203,6 +204,55 @@ impl MessageLog {
...
+    /// zero.
+    ///
+    /// Note that the first segment yielded by the iterator may contain messages
+    /// with an offset _smaller_ than the argument, as segments do not currently
+    /// support slicing.
+    ///
+    /// The iterator represents a _snapshot_ of the log at the time this method
+    /// is called. That is, segments created after the method returns will not
+    /// appear in the iteration. The last segment yielded by the iterator may be
+    /// incomplete (i.e. still be appended to).
```
Comment (cloutiertyler) [inline crates/core/src/db/message_log.rs:233]: These doc comments are A+ 🙏

File: crates/core/src/db/commit_log.rs [hunk 2]
```diff
@@ -121,3 +127,150 @@ impl CommitLog {
         }
     }
 }
+
+/// A read-only view of a [`CommitLog`].
+pub struct CommitLogView {
+    mlog: Option<Arc<Mutex<MessageLog>>>,
```
Comment (cloutiertyler) [inline crates/core/src/db/commit_log.rs:133]: I might suggest implementing the `CommitLog` by giving it a `CommitLogView` internally, since it does already share the same fields. Hmm that doesn't seem like the right solution either, but I wonder if we can have a clearer abstraction here.

It seems like we are almost trying to create the concept of a "File" (i.e. `CommitLogView`) and a "BufferedFile" (i.e. `CommitLog`) where the buffered thing is unwritten commit which is only periodically flushed. I wonder if something like that might fit better down the road.

File: crates/core/src/db/message_log.rs [hunk 3]
```diff
@@ -217,6 +267,124 @@ impl MessageLog {
     }
 }
 
+/// A read-only view of an on-disk [`Segment`] of the [`MessageLog`].
+///
+/// The underlying file is opened lazily when calling [`SegmentView::try_into_iter`]
+/// or [`SegmentView::try_into_file`].
+#[derive(Clone, Debug)]
+pub struct SegmentView {
```
Comment (cloutiertyler) [inline crates/core/src/db/message_log.rs:275]: Why is this required instead of `MessageLogIter`? Also presumably we should implement the internals of `MessageLogIter` with `SegementView` because we're duplicating some code that really needs to stay in sync.

**Review by cloutiertyler — COMMENTED** [review]
More or less seems okay, although I have some important questions inline.

File: crates/core/src/db/commit_log.rs (hunk 2 above)
Comment (kim) [reply crates/core/src/db/commit_log.rs:133]: The purpose of `CommitLogView` is for `RelationalDB` to hand out a view over its _internal state_ in order to serve up log segments. It must hide `append`, but changing the visibility of `append` isn't helpful.

`CommitLogView` could be thought of as a newtype around `CommitLog`, but internally that'd be copying more data than strictly required. Perhaps if there was an `Arc` around the whole `CommitLog`, but I'm always wary about contention.

File: crates/core/src/db/message_log.rs (hunk 3 above)
Comment (kim) [reply crates/core/src/db/message_log.rs:275]: I was going to ask about `MessageLogIter`, and perhaps submit a separate patch: the issue is, it has very different semantics. For one, it panics on I/O errors, which isn't great when it's not a matter of database consistency. More importantly, though, `segments_for_offset` returns the **first** segment if `offset` does not fall into the **last** segment -- the significance of which I don't understand.

So imho this would require to write a number of tests at least, which I wouldn't want to conflate with this patch (which does not change any database internals).

Comment (kim) [reply crates/core/src/db/message_log.rs:275]: > Why is this required instead of MessageLogIter?

It is also not helpful to borrow `MessageLog` when that value is behind a mutex.

**Review by cloutiertyler — APPROVED** [review]
LGTM ultimately, although I'd like to think about a different philosophy on this
